ackermann_steering_controller/test/test_ackermann_steering_controller.cpp

```diff
@@ -0,0 +1,294 @@
+    COMMON_THRESHOLD);
+  EXPECT_NEAR(
+    controller_->command_interfaces_[STATE_TRACTION_LEFT_WHEEL].get_value(), 0.22222222222222224,
+    COMMON_THRESHOLD);
+  EXPECT_NEAR(
+    controller_->command_interfaces_[STATE_STEER_RIGHT_WHEEL].get_value(), 1.4179821977774734,
+    COMMON_THRESHOLD);
+  EXPECT_NEAR(
+    controller_->command_interfaces_[STATE_STEER_LEFT_WHEEL].get_value(), 1.4179821977774734,
+    COMMON_THRESHOLD);
+
+  EXPECT_TRUE(std::isnan((*(controller_->input_ref_.readFromRT()))->twist.linear.x));
+  EXPECT_EQ(controller_->reference_interfaces_.size(), joint_reference_interfaces_.size());
+  for (const auto & interface : controller_->reference_interfaces_)
+  {
+    EXPECT_TRUE(std::isnan(interface));
+  }
+}
+
+TEST_F(AckermannSteeringControllerTest, receive_message_and_publish_updated_status)
+{
+  SetUpController();
+  rclcpp::executors::MultiThreadedExecutor executor;
+  executor.add_node(controller_->get_node()->get_node_base_interface());
+
+  ASSERT_EQ(controller_->on_configure(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  ASSERT_EQ(controller_->on_activate(rclcpp_lifecycle::State()), NODE_SUCCESS);
+
+  ASSERT_EQ(
+    controller_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+
+  ControllerStateMsg msg;
+  subscribe_and_get_messages(msg);
+
+  EXPECT_EQ(msg.linear_velocity_command.data[STATE_TRACTION_RIGHT_WHEEL], 1.1);
+  EXPECT_EQ(msg.linear_velocity_command.data[STATE_TRACTION_LEFT_WHEEL], 3.3);
+  EXPECT_EQ(msg.steering_angle_command.data[0], 2.2);
+  EXPECT_EQ(msg.steering_angle_command.data[1], 4.4);
+
+  publish_commands();
+  ASSERT_TRUE(controller_->wait_for_commands(executor));
+
+  ASSERT_EQ(
+    controller_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+
+  EXPECT_NEAR(
+    controller_->command_interfaces_[CMD_TRACTION_RIGHT_WHEEL].get_value(), 0.22222222222222224,
+    COMMON_THRESHOLD);
+  EXPECT_NEAR(
+    controller_->command_interfaces_[CMD_TRACTION_LEFT_WHEEL].get_value(), 0.22222222222222224,
+    COMMON_THRESHOLD);
+  EXPECT_NEAR(
+    controller_->command_interfaces_[CMD_STEER_RIGHT_WHEEL].get_value(), 1.4179821977774734,
+    COMMON_THRESHOLD);
+  EXPECT_NEAR(
+    controller_->command_interfaces_[CMD_STEER_LEFT_WHEEL].get_value(), 1.4179821977774734,
+    COMMON_THRESHOLD);
+
+  subscribe_and_get_messages(msg);
+
+  EXPECT_NEAR(
+    msg.linear_velocity_command.data[CMD_TRACTION_RIGHT_WHEEL], 0.22222222222222224,
+    COMMON_THRESHOLD);
+  EXPECT_NEAR(
+    msg.linear_velocity_command.data[CMD_TRACTION_LEFT_WHEEL], 0.22222222222222224,
+    COMMON_THRESHOLD);
+  EXPECT_NEAR(msg.steering_angle_command.data[0], 1.4179821977774734, COMMON_THRESHOLD);
+  EXPECT_NEAR(msg.steering_angle_command.data[1], 1.4179821977774734, COMMON_THRESHOLD);
+}
+
+int main(int argc, char ** argv)
```

Comment:
Not really. We have to have this because we are using an integrated test with generate_parameter_library. This enables us to have exemplary parameter sets “on board” that confirm functionality, and one can simply copy those files and adjust them. I think this is a neat end-user life improvement having such a thing. Also, one can simply put their parameters into tests to check the behavior if needed – good for debugging so strange setups.

Reply:
This piece of code has none of what you mention. Please elaborate.